### PR TITLE
Sensor options panel within the project properties dialog

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR builds on top of the core sensors framework (https://github.com/qgis/QGIS/pull/52297) and adds the first part of the GUI interface: a sensor table widget that displays registered sensors within a project with actions to connect and disconnect from sensors. The table is accessible via a new sensors panel in the project properties dialog:

image

To avoid a large PR, I'll open another one after this is reviewed covering sensor creation + configuration GUI (see https://github.com/qgis/QGIS/pull/52376)

_Sponsored by [Sevenson Environmental Services](https://sevenson.com/)._

### CODE REVIEW

File: src/core/sensor/qgssensormodel.cpp [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+  {
+    case Qt::DisplayRole:
+    case Qt::ToolTipRole:
+    case Qt::EditRole:
+    {
+      switch ( index.column() )
+      {
+        case Name:
+        {
+          QgsAbstractSensor *sensor = mSensorManager->sensor( mSensorIds[index.row()] );
```
Comment (nyalldawson) [inline src/core/sensor/qgssensormodel.cpp:50]: I'd move this up to sit before the switch, with a single `if ( !sensor ) return QVariant();` before we handle the role/columns.

File: src/core/sensor/qgssensormodel.cpp [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+    case SensorType:
+    {
+      QgsAbstractSensor *sensor = mSensorManager->sensor( mSensorIds[index.row()] );
+      return sensor ? sensor->type() : QString();
+    }
+
+    case SensorId:
+    {
+      return mSensorIds[index.row()];
```
Comment (nyalldawson) [inline src/core/sensor/qgssensormodel.cpp:89]: potential crash here if index.row() isn't present in the map -- would be solved by the above comment though.

File: src/core/sensor/qgssensormodel.cpp [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    {
+      return QVariant::fromValue<QgsAbstractSensor *>( mSensorManager->sensor( mSensorIds[index.row()] ) );
+    }
+
+    default:
+      return QVariant();
+  }
+#ifndef _MSC_VER // avoid warning
+  return QVariant();  // avoid warning
+#endif
```
Comment (nyalldawson) [inline src/core/sensor/qgssensormodel.cpp:126]: BUILTIN_UNREACHABLE

File: src/core/sensor/qgssensormodel.cpp [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+  {
+    case Name:
+    {
+      QgsAbstractSensor *sensor = mSensorManager->sensor( mSensorIds[index.row()] );
+      if ( sensor )
+      {
+        sensor->setName( value.toString() );
+        return true;
+      }
+    }
```
Comment (nyalldawson) [inline src/core/sensor/qgssensormodel.cpp:144]: ` 
   else
   {
       break;
    }
  
default:
   break;
`

File: src/core/sensor/qgssensormodel.cpp [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+  }
+
+  return false;
+}
+
+Qt::ItemFlags QgsSensorModel::flags( const QModelIndex &index ) const
+{
+  Qt::ItemFlags flags = QAbstractItemModel::flags( index );
+  if ( index.isValid() && index.column() == Name )
```
Comment (nyalldawson) [inline src/core/sensor/qgssensormodel.cpp:153]: should this also check if the row matches a sensor entry first?

File: src/core/sensor/qgssensormodel.cpp [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+  }
+}
+
+QVariant QgsSensorModel::headerData( int section, Qt::Orientation orientation, int role ) const
+{
+  if ( role == Qt::DisplayRole )
+  {
+    if ( orientation == Qt::Vertical ) //row
+    {
+      return QVariant( section );
```
Comment (nyalldawson) [inline src/core/sensor/qgssensormodel.cpp:169]: section + 1 ?

File: src/core/sensor/qgssensormodel.cpp [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    }
+    else
+    {
+      switch ( section )
+      {
+        case Name:
+          return QVariant( tr( "Name" ) );
+
+        case LastValue:
+          return QVariant( tr( "Last value" ) );
```
Comment (nyalldawson) [inline src/core/sensor/qgssensormodel.cpp:179]: Will need double-checking, but pretty sure our convention is for Title Case for these headers
```suggestion
          return QVariant( tr( "Last Value" ) );
```

File: src/core/sensor/qgssensormodel.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ * \since QGIS 3.32
+ */
+class CORE_EXPORT QgsSensorModel: public QAbstractItemModel
+{
+    Q_OBJECT
+
+  public:
+
+    //! Model columns
+    enum Column
```
Comment (nyalldawson) [inline src/core/sensor/qgssensormodel.h:39]: enum class

**Comment by nyalldawson** [comment]
@nirvn the code is looking good, but I'm not convinced a dock widget is the right approach here. I think it should instead be a tab in the Project Properties dialog, since it's all project-level configuration.

If the live readout is desirable in a non-blocking dock then I'd implement that as a plugin.

**Comment by nirvn** [comment]
@nyalldawson , review addressed. I've moved the sensors table within a new sensors panel in the project properties. We can revisit the idea of a sensors dock (or another form of UI/UX within the main window) to handle sensor activation/deactivation and value later.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**